src/rgen/Microsoft.Macios.Generator/BindingSourceGeneratorGenerator.cs

```diff
@@ -93,8 +85,85 @@ static void AddPipeline<T> (IncrementalGeneratorInitializationContext context) w
 
 		var codeChanges = CodeChanges.FromDeclaration (context.SemanticModel, declarationSyntax);
 		// if code changes are null, return the default value and a false to later ignore the change
-		return codeChanges is not null ?
-			(codeChanges.Value, isBindingType) : (default, false);
+		return codeChanges is not null ? (codeChanges.Value, isBindingType) : (default, false);
+	}
+
+	static void GenerateCode (SourceProductionContext context, Compilation compilation,
+		ImmutableArray<CodeChanges> changesList)
+	{
+		// the process is as follows, get all the changes we have received from the incremental generator,
+		// loop over them and based on the CodeChange.BindingType we are going to build the symbol context
+		// and emitter. Those are later used to generate the code.
+		//
+		// Once all the enums, classes and interfaces have been processed, we will use the data collected
+		// in the RootBindingContext to generate the library and trampoline code.
+		var rootContext = new RootBindingContext (compilation);
+		foreach (var change in changesList) {
+			var semanticModel = compilation.GetSemanticModel (change.SymbolDeclaration.SyntaxTree);
+			// This is a bug in the roslyn analyzer for roslyn generator https://github.com/dotnet/roslyn-analyzers/issues/7436
+#pragma warning disable RS1039
+			if (semanticModel.GetDeclaredSymbol (change.SymbolDeclaration) is not INamedTypeSymbol namedTypeSymbol)
+#pragma warning restore RS1039
+				continue;
+
+			// init sb and add the header
+			var sb = new TabbedStringBuilder (new ());
+			sb.WriteHeader ();
+			if (EmitterFactory.TryCreate (change, rootContext, semanticModel, namedTypeSymbol, sb, out var emitter)) {
+				// write the using statements
+				CollectUsingStatements (change.SymbolDeclaration.SyntaxTree, sb, emitter);
+
+				if (emitter.TryEmit (out var diagnostics)) {
+					// only add file when we do generate code
+					var code = sb.ToString ();
+					context.AddSource ($"{emitter.SymbolNamespace}/{emitter.SymbolName}.g.cs",
+						SourceText.From (code, Encoding.UTF8));
+				} else {
+					// add to the diagnostics and continue to the next possible candidate
+					foreach (Diagnostic diagnostic in diagnostics) {
+						context.ReportDiagnostic (diagnostic);
+					}
+				}
+			} else {
+				// we don't have a emitter for this type, so we can't generate the code, add a diagnostic letting the
+				// user we do not support what he is trying to do
+				// we don't have a emitter for this type, so we can't generate the code, add a diagnostic letting the
+				// user we do not support what he is trying to do
+				context.ReportDiagnostic (Diagnostic.Create (RBI0000,
+					change.SymbolDeclaration.GetLocation (),
+					namedTypeSymbol.ToDisplayString ().Trim ()));
```

Comment:
Would it be possible to always add the actual diagnostic message as a comment? It makes it much easier to see that the message is correct and makes sense when it's with context that shows how it's used (as opposed to in the resx file, which has no context), and it also makes it easier to grep for error messages and find the correct source code location right away:

```suggestion
				context.ReportDiagnostic (Diagnostic.Create (RBI0000, // An unexpected error ocurred while processing '{0}'. Please fill a bug report at https://github.com/xamarin/xamarin-macios/issues/new.
					change.SymbolDeclaration.GetLocation (),
					namedTypeSymbol.ToDisplayString ().Trim ()));
```

Although this one in particular (RBI0000) is probably going to be common enough that an extension method on `GeneratorSyntaxContext` might be nice:

```cs
public static void ReportUnexpectedError (this GeneratorSyntaxContext self, object location, string? symbol)
{
	self.ReportDiagnostic (Diagnostic.Create (
		RBI0000, // An unexpected error ocurred while processing '{0}'. Please fill a bug report at https://github.com/xamarin/xamarin-macios/issues/new.
		location,
		symbol));
}
```

Reply:
What about moving it to a static class with all of them and add documentation so that when you hover over the issue you get the message, for example:

![Screenshot 2024-11-01 at 5 49 05 PM](https://github.com/user-attachments/assets/21a45abd-45ad-47b3-8d5d-9faecdaa03dd)

This guarantees that you get always the context of the message. And then you can do a Find usages or the similar in the IDE.

Reply:
> What about moving it to a static class with all of them and add documentation

I like this on a general principle, but it doesn't solve my use case:

> when you hover over the issue you get the message

I usually use SublimeText... also grep doesn't work with this approach either, nor do code reviews in GitHub.